### PR TITLE
[MRG+1] Doc review guidelines

### PR DESCRIPTION
**Pull request by tacaswell**
This is my understanding of the current review guide lines and a starting point to if we want to formalize it any more.

We clearly have a problem getting PRs reviewed and merged in a timely fashion.

attn @matplotlib/developers

### CODE REVIEW

File: doc/devel/coding_guide.rst [hunk 1]
```diff
@@ -165,6 +165,48 @@ C/C++ extensions
...
+
+
+Backports
+=========
+
+
+When doing back ports please include the branch you backported the
+commit to along with the SHA in a comment on the original PR.
+
+A quick guide to how to back port a merge commit::
```
Comment (story645) [inline doc/devel/coding_guide.rst]: typo: backports,  not back ports
and `when doing backports, please include..."

File: doc/devel/coding_guide.rst [hunk 2]
```diff
@@ -165,6 +165,48 @@ C/C++ extensions
...
+  git checkout target_branch
+  git merge --ff-only matplotlib/target_branch
+  git cherry-pick -m 1 TARGET_SHA
+  gitk   # to look at it
+  # local tests? (use your judgment)
+  git push DANGER target_branch
+  # leave a comment on PR noting sha of the resulting commit
+  # from the cherry-pick + branch it was moved to
+
+here matplotlib is a read-only remote to the matplotlib/matplotlib repo and DANGER in a read/write remote to the matplotlib/matplotlib repo.
```
Comment (story645) [inline doc/devel/coding_guide.rst]: typo? I think DANGER is and not DANGER in...otherwise it doesn't quite make sense.

File: doc/devel/coding_guide.rst [hunk 3]
```diff
@@ -165,6 +165,48 @@ C/C++ extensions
...
 
 
+PR Review guidelines
+====================
+
+* We run on a 'use your judgement' system (for better or worse) and do
+  not have a strict nose-count rule.  If you have a commit bit, then
+  you are trusted to use it and to know when to ask for more review.
+
+* Do not self merge with the exception of un-breaking travis.
```
Comment (QuLogic) [inline doc/devel/coding_guide.rst]: Travis-CI; but what about AppVeyor?

File: doc/devel/coding_guide.rst [hunk 4]
```diff
@@ -165,6 +165,48 @@ C/C++ extensions
 
 
 
+PR Review guidelines
+====================
+
+* We run on a 'use your judgement' system (for better or worse) and do
+  not have a strict nose-count rule.  If you have a commit bit, then
```
Comment (story645) [inline doc/devel/coding_guide.rst]: since we're switching to pylint, nose-count should likely be changed to something...dunno what. And what's a commit bit?

Comment (QuLogic) [inline doc/devel/coding_guide.rst]: I think by "nose-count", @tacaswell meant we don't have a minimum number of reviewers before a PR can be merged and not a reference to nosetests, but it's a bit of an idiom.

Commit bit means membership in the developer role for the repo (generally.)

File: doc/devel/coding_guide.rst (hunk 2 above)
Comment (anntzer) [inline doc/devel/coding_guide.rst]: Also, how do you create a readonly remote?  I'm always scared of accidentally pushing to the main repo.

Comment (NelleV) [inline doc/devel/coding_guide.rst]: @anntzer The easiest solution is to use the https remote url. This way, you are always prompt for your username and password. It doesn't prevent you from actually pushing, but should raise red flags.

File: doc/devel/coding_guide.rst (hunk 4 above)
Comment (NelleV) [inline doc/devel/coding_guide.rst]: @tacaswell I'd switch to a 2 core dev validation, with explicit approval for merge except for documentation changes.
Else, we'll stay in the situation we are in, where everyone comments and no one merges.

Comment (tacaswell) [inline doc/devel/coding_guide.rst]: Being a bit too glib/jargony.  'nose-count' in X devs signing off on it and 'commit bit' is write access the official repositories.

File: doc/devel/coding_guide.rst (hunk 3 above)
Comment (tacaswell) [inline doc/devel/coding_guide.rst]: again, being too glib

File: doc/devel/coding_guide.rst (hunk 2 above)
Comment (NelleV) [inline doc/devel/coding_guide.rst]: @anntzer else, you can use this command: `git remote set-url matplotlib --push "read-only"`

Comment (QuLogic) [inline doc/devel/coding_guide.rst]: Set an empty/invalid push url with `git remote set-url --push <remote> blah-blah`.

Comment (anntzer) [inline doc/devel/coding_guide.rst]: Probably worth mentioning this somewhere in the dev docs?

File: doc/devel/coding_guide.rst (hunk 4 above)
Comment (story645) [inline doc/devel/coding_guide.rst]: I like @NelleV's suggestion, possibly with the addition of github flags or something ("+1") so it's easier to find the ready to merge PRs.

File: doc/devel/coding_guide.rst (hunk 2 above)
Comment (Kojoley) [inline doc/devel/coding_guide.rst]: @NelleV if you are using [git credentials manager](https://git-scm.com/book/be/v2/Git-Tools-Credential-Storage) you will not be asked for the login/password on a push.

Comment (NelleV) [inline doc/devel/coding_guide.rst]: @Kojoley I use ssh keys. I don't have a valid reason to use https over ssh as all the ports are open in the institutes I work in.

Comment (Kojoley) [inline doc/devel/coding_guide.rst]: > Also, how do you create a readonly remote?
> @anntzer The easiest solution is to use the https remote url.

My comment was related to this ^.

File: doc/devel/coding_guide.rst (hunk 4 above)
Comment (tacaswell) [inline doc/devel/coding_guide.rst]: It is somewhat counter intuitive that going to a '2 devs must sign off' scheme would increase the rate at which things get merged, but I suspect it might be true.

File: doc/devel/coding_guide.rst [hunk 5]
```diff
@@ -183,6 +183,30 @@ PR Review guidelines
...
+Backports
+=========
+
+
+When doing back ports please include the branch you backported the
+commit to along with the SHA in a comment on the original PR.
+
+A quick guide to how to back port a merge commit::
+
+  git remote update
```
Comment (dopplershift) [inline doc/devel/coding_guide.rst]: Well I just learned a new command...

File: doc/devel/coding_guide.rst (hunk 2 above)
Comment (tacaswell) [inline doc/devel/coding_guide.rst]: This is a sub-section of my `git remote -v`

```
DANGER  git@github.com:matplotlib/matplotlib.git (fetch)
DANGER  git@github.com:matplotlib/matplotlib.git (push)
matplotlib      git://github.com/matplotlib/matplotlib.git (fetch)
matplotlib      git://github.com/matplotlib/matplotlib.git (push)
tacaswell       git@github.com:tacaswell/matplotlib.git (fetch)
tacaswell       git@github.com:tacaswell/matplotlib.git (push)
```

I also use ssh keys on machines where I can (I also have 2FA turned on so https requires an oauth token) so I can push to `tacaswel` and `DANGER`, but not `matplotlib`. 

```
$ git push matplotlib v2.x
fatal: remote error: 
  You can't push to git://github.com/matplotlib/matplotlib.git
  Use https://github.com/matplotlib/matplotlib.git
```

 Due to the fear of pushing to the wrong remote (which I have done) I am also very pedantic about being explicit about remotes when pushing and have git turned down to the least magical 'guess what you mean by `git push`'.  

Having to type the word 'DANGER' in all caps (or at least see it in magit) does remind you you are about to do something dangerous ;) 

I got this naming idea from @dmcdougall

File: doc/devel/coding_guide.rst (hunk 4 above)
Comment (NelleV) [inline doc/devel/coding_guide.rst]: If it is shown in the title bar, it is also a quick way to scan for PRs that are lacking one review.

Comment (dopplershift) [inline doc/devel/coding_guide.rst]: Well, it removes the roadblock of wondering whether something is ready....and shifts it to devs being willing to formally sign off.

IMO, the new formal PR review stuff would make it much easier to track who's signed off.

Comment (tacaswell) [inline doc/devel/coding_guide.rst]: have skbeam / sklearn also looked at doing this with labels?   We do have waffle.io set up which gives some 'auto label' management by moving things between columns.  Could have 3 review related columns 'needs revision', 'needs first review', 'needs final review'.  But that might be getting to fancy.

Comment (story645) [inline doc/devel/coding_guide.rst]: the only reason I don't favor the title bar is 'cause I read [MRG+1] as 1 commit ahead of merged code, which wrong but a habit my brain won't break.

Comment (QuLogic) [inline doc/devel/coding_guide.rst]: The formal PR review stuff still has a few kinks to work out; outdated comments never disappear, approval is "sticky" even if the submitter changes the PR substantially, etc.

Doesn't waffle auto-label new stuff as "needs review" already? There's definitely "needs review" and "needs revision" columns there now. Or did auto-labelling get disabled along with wafflebot?

Comment (NelleV) [inline doc/devel/coding_guide.rst]: You can edit the PR title before merging.
I personally don't like labels. They are hard to filter, clutter the UI, and half of them are IMO useless. They are also slower to edit (to move from MRG+1 to MRG+2 would require to remove a label and add a new one, or keep the two, but that's a lot of useless information displayed to the user).

Stéfan van der Walt (one of the main maintainers of scikit-image) recently strongly advocated the new projects features of github for classifying PRs into different categories. I haven't used it yet, but I can give feedback in a month or so once I'll use it on common projects with him.

The scikit-learn team is sticking to editing the PR title as we found it is, right now, the most efficient way to scan quickly through PRs to have an overview of the WIP (work in progress), and the different review states of PRs. But then our commit logs look like this: https://github.com/scikit-learn/scikit-learn/commits/master (which I am personally fine with)

Comment (story645) [inline doc/devel/coding_guide.rst]: if we go the title route, can that please be documented somewhere? (Possibly in this doc). I know some of it is borrowed from [scipy](http://docs.scipy.org/doc/numpy-dev/dev/gitwash/development_workflow.html) but even a link to that would be helpful.

Comment (story645) [inline doc/devel/coding_guide.rst]: @efiring what about reforming the labels so that they follow something like https://github.com/numpy/numpy/labels? I'm a huge fan of labels for quickly sorting out the things I can't do.

Comment (tacaswell) [inline doc/devel/coding_guide.rst]: It appears that we lost what waffle stuff we had when wafflebot got turned off.  I poked at it a bit and the events it will respond to is very limited and it looks like the new review tools do not have public web-hooks yet (so in principle we could auto-flag PRs when someone does an 'approve' review).

I think automatic labels > 'hand edited' titles > hand edited labels.

File: doc/devel/coding_guide.rst [hunk 6]
```diff
@@ -165,6 +165,64 @@ C/C++ extensions
...
+* Squashing is case-by-case.  The balance is between burden on the
+  contributor, keeping a relatively clean history, and keeping a
+  history usable for bisecting.  The only time we are really strict
+  about it is to eliminate binary files (ex multiple test image
+  re-generations) and to remove upstream merges.
+
+* Be patient with new contributors.
+
+* Do not let perfect be the enemy of the good, particularly for
+  documentation or example PRs.  If you find your self making many
```
Comment (NelleV) [inline doc/devel/coding_guide.rst]: your self -> yourself ?

File: doc/devel/coding_guide.rst [hunk 7]
```diff
@@ -165,6 +165,64 @@ C/C++ extensions
...
+PR Review guidelines
+====================
+
+* If you have a commit bit, then you are trusted to use it.  Please
+  help review and merge PRs!
+
+* Two developers (those with commit rights) should review all pull
+  requests.  If you are the first to review a PR please and approve of
+  the changes please edit the title to include ``'[MRG+1]'`` and use
+  the github 'approve review' tool to mark it as such.  If you are a
```
Comment (mdboom) [inline doc/devel/coding_guide.rst]: Link to Github's documentation of the approve review tool.

**Comment by NelleV** [comment]
@mdboom Just a comment: thought it actually doesn't block merging the pull request, the new approve changes/request changes of github need to be "removed" by hand. It is not because @tacaswell updated his PR that the status will be removed.

**Comment by NelleV** [comment]
Thanks @tacaswell for the documentation!
I've created a ticket following the remarks on the cherry-picking section.
